common/nym-lp/src/session.rs

```diff
@@ -0,0 +1,1898 @@
+                &local_ed25519_keypair.1.to_bytes(),
+            )
+            .expect("Valid ed25519 public key"),
+            remote_ed25519_public: ed25519::PublicKey::from_bytes(&remote_ed25519_key.to_bytes())
+                .expect("Valid ed25519 public key"),
+            local_x25519_private: local_x25519_key.clone(),
+            remote_x25519_public: remote_x25519_key.clone(),
+            salt: *salt,
+        })
+    }
+
+    pub fn next_packet(&self, message: LpMessage) -> Result<LpPacket, LpError> {
+        let counter = self.next_counter();
+        let header = LpHeader::new(self.id(), counter);
+        let packet = LpPacket::new(header, message);
+        Ok(packet)
+    }
+
+    /// Generates the next counter value for outgoing packets.
+    pub fn next_counter(&self) -> u64 {
+        self.sending_counter.fetch_add(1, Ordering::Relaxed)
+    }
+
+    /// Performs a quick validation check for an incoming packet counter.
+    ///
+    /// This should be called before performing any expensive operations like
+    /// decryption/Noise processing to efficiently filter out potential replay attacks.
+    ///
+    /// # Arguments
+    ///
+    /// * `counter` - The counter value to check
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the counter is likely valid
+    /// * `Err(LpError::Replay)` if the counter is invalid or a potential replay
+    pub fn receiving_counter_quick_check(&self, counter: u64) -> Result<(), LpError> {
+        // Branchless implementation uses SIMD when available for constant-time
+        // operations, preventing timing attacks. Check before crypto to save CPU cycles.
+        let counter_validator = self.receiving_counter.lock();
+        counter_validator
+            .will_accept_branchless(counter)
+            .map_err(LpError::Replay)
+    }
+
+    /// Marks a counter as received after successful packet processing.
+    ///
+    /// This should be called after a packet has been successfully decoded and processed
+    /// (including Noise decryption/handshake step) to update the replay protection state.
+    ///
+    /// # Arguments
+    ///
+    /// * `counter` - The counter value to mark as received
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the counter was successfully marked
+    /// * `Err(LpError::Replay)` if the counter cannot be marked (duplicate, too old, etc.)
+    pub fn receiving_counter_mark(&self, counter: u64) -> Result<(), LpError> {
+        let mut counter_validator = self.receiving_counter.lock();
+        counter_validator
+            .mark_did_receive_branchless(counter)
+            .map_err(LpError::Replay)
+    }
+
+    /// Returns current packet statistics for monitoring.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * The next expected counter value for incoming packets
+    /// * The total number of received packets
+    pub fn current_packet_cnt(&self) -> (u64, u64) {
```

Comment:
nit: I think it'd be more readable (and less error prone) if that tuple was wrapped in a struct. it's non-obvious and easy to confuse `.0` and `.1` values